### PR TITLE
Handle dns ovh 

### PR DESCRIPTION
**Pull request by guldil**
Hello,

Like Gandi DNS, i add the OVH DNS.

First step, create a token for API : https://eu.api.ovh.com/createToken/

in config.yml :
````yaml
certificates:
  - domain: www.domain.com
    solver:
      name: ovh
      app_key: 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'
      app_secret: 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'
      consumer_key: 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'
      end_point: 'ovh-eu'
````
There are more steps to update DNS with API than Gandi 

- you have to POST TXT then Refresh Zone 
- for DELETE you have to get the ID then DELETE and refresh zone

It's working i test single domain and wildcard domain.

### CODE REVIEW

**Review by tgalopin — COMMENTED** [review]
Hi! Thanks for the proposal!

Unfortunately, there are quite big problems with this PR: you shouldn't include your IDE Git files, you should try to modify the least amount fo code possible (no need to update composer.json for instance) and finally you should add tests.

If you wish to finalize it, I can do a more precise review, but it's going to take some work so I prefer t ask before: do you wish to finalize, including creating tests? If so I'll review :)

**Comment by guldil** [comment]
Hello,

No problem, I will redo a PR from version 1.3.0 taking into account your remarks.

Guldil

File: .gitignore [hunk 1]
```diff
@@ -5,3 +5,6 @@ vendor/
 .php_cs.cache
 doc/.couscous
 .subsplit
+/config.yml
+/.buildpath
+/.project
```
Comment (tgalopin) [inline .gitignore:10]: You should use a global gitignore for your personal tools: https://gist.github.com/subfuzion/db7f57fff2fb6998a16c

File: .settings/.gitignore [hunk 2]
```diff
@@ -0,0 +1,2 @@
+/org.eclipse.php.core.prefs
```
Comment (tgalopin) [inline .settings/.gitignore:1]: Same for this file

File: composer.json [hunk 3]
```diff
@@ -1,91 +1,90 @@
 {
-    "name": "acmephp/acmephp",
```
Comment (tgalopin) [inline composer.json:2]: You should only add the new package line: updating this whole file makes reviewing difficult

File: src/Core/Challenge/Dns/OvhSolver.php [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+/**
+ * ACME DNS solver with automate configuration of a Gandi.Net.
+ *
+ * @author Alexander Obuhovich <aik.bold@gmail.com>
+ */
+class OvhSolver implements MultipleChallengesSolverInterface, ConfigurableServiceInterface
+{
+    use LoggerAwareTrait;
+    /**
```
Comment (tgalopin) [inline src/Core/Challenge/Dns/OvhSolver.php:32]: Missing empty line before

File: src/Core/Challenge/Dns/OvhSolver.php [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+     * @var string
+     */
+    private $domainDelegation;
+
+    /**
+     * @param DnsDataExtractor $extractor
+     * @param ClientInterface  $client
+     */
+    public function __construct(
+        DnsDataExtractor $extractor = null,
```
Comment (tgalopin) [inline src/Core/Challenge/Dns/OvhSolver.php:77]: You should be able to put the 2 arguments on the same line

File: src/Core/Challenge/Dns/OvhSolver.php [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    public function solveAll(array $authorizationChallenges)
+    {
+        Assert::allIsInstanceOf($authorizationChallenges, AuthorizationChallenge::class);
+
+        foreach ($authorizationChallenges as $authorizationChallenge) {
+            $this->logger->debug('Create TXT for '.$authorizationChallenge->getDomain());
```
Comment (tgalopin) [inline src/Core/Challenge/Dns/OvhSolver.php:123]: Creating TXT record for {domain} on OVH

File: src/Core/Challenge/Dns/OvhSolver.php [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+
+            $topLevelDomain = $this->getTopLevelDomain($authorizationChallenge->getDomain());
+            $recordName = $this->extractor->getRecordName($authorizationChallenge);
+            $recordValue = $this->extractor->getRecordValue($authorizationChallenge);
+
+            $subDomain = \str_replace('.'.$topLevelDomain.'.', '', $recordName);
+
+            $client = new Client(['http_errors' => false]);
+
+            $ovh = new Api($this->appKey,
```
Comment (tgalopin) [inline src/Core/Challenge/Dns/OvhSolver.php:133]: The coding style here seems weird: you can either put everything on the same line, or all arguments on a new line

File: src/Core/Challenge/Dns/OvhSolver.php [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+
+            $result = $ovh->post('/domain/zone/'.$topLevelDomain.'/record', [
+            'fieldType' => 'TXT',
+            'subDomain' => $subDomain,
+            'target' => $recordValue,
+            'ttl' => 600,
+            ]);
+
+            if (isset($result['message']) && '' !== $result['message']) {
+                $this->logger->error('OVH Exception Post = '.$result['message'].' '.$topLevelDomain);
```
Comment (tgalopin) [inline src/Core/Challenge/Dns/OvhSolver.php:151]: You probably don't need to use the logger here as the monitoring tools will catch the exception

File: src/Core/Challenge/Dns/OvhSolver.php [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+                throw new \Exception('OVH Exception Post = '.$result['message'].' '.$topLevelDomain);
```
Comment (tgalopin) [inline src/Core/Challenge/Dns/OvhSolver.php:152]: An error occurred while creating the TXT record for {domain} on OVH: {message}

Tip: you can use [sprintf](https://www.php.net/manual/fr/function.sprintf.php) to ease concatenation

File: src/Core/Challenge/Dns/OvhSolver.php [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+            }
+
+            $this->logger->info('Refresh zone for '.$authorizationChallenge->getDomain());
```
Comment (tgalopin) [inline src/Core/Challenge/Dns/OvhSolver.php:155]: Refreshing

File: src/Core/Challenge/Dns/OvhSolver.php [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+            $result = $ovh->post('/domain/zone/'.$topLevelDomain.'/refresh');
+
+            if (isset($result['message']) && '' !== $result['message']) {
+                $this->logger->error('OVH Exception Refresh= '.$result['message'].' '.$topLevelDomain);
```
Comment (tgalopin) [inline src/Core/Challenge/Dns/OvhSolver.php:159]: Same here: not necessary

File: src/Core/Challenge/Dns/OvhSolver.php [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+                throw new \Exception('OVH Exception Refresh= '.$result['message'].' '.$topLevelDomain);
```
Comment (tgalopin) [inline src/Core/Challenge/Dns/OvhSolver.php:160]: An error occurred while refreshing the DNS zone for {domain} on OVH: {message}

File: src/Core/Challenge/Dns/OvhSolver.php [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    public function cleanupAll(array $authorizationChallenges)
+    {
+        Assert::allIsInstanceOf($authorizationChallenges, AuthorizationChallenge::class);
+
+        foreach ($authorizationChallenges as $authorizationChallenge) {
+            $this->logger->debug('Cleanup record id for '.$authorizationChallenge->getDomain());
```
Comment (tgalopin) [inline src/Core/Challenge/Dns/OvhSolver.php:183]: Cleaning TXT record for {domain} on OVH

File: src/Core/Challenge/Dns/OvhSolver.php [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+                $topLevelDomain = $this->domainDelegation;
+            }
+
+            $result = $ovh->get('/domain/zone/'.$topLevelDomain.'/record', [
+            'fieldType' => 'TXT', // Filter the value of fieldType property (like) (type: zone.NamedResolutionFieldTypeEnum)
+            'subDomain' => $subDomain, // Filter the value of subDomain property (like) (type: string)
+            ]);
+
+            if (isset($result['message']) && '' !== $result['message']) {
+                $this->logger->error('OVH Exception Get= '.$result['message'].' '.$topLevelDomain);
```
Comment (tgalopin) [inline src/Core/Challenge/Dns/OvhSolver.php:208]: Same here

File: src/Core/Challenge/Dns/OvhSolver.php [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+                throw new \Exception('OVH Exception Get= '.$result['message'].' '.$topLevelDomain);
```
Comment (tgalopin) [inline src/Core/Challenge/Dns/OvhSolver.php:209]: You can adapt the message I proposed in previous comments here and in the other exceptions

**Review by tgalopin — COMMENTED** [review]
Hi! Awesome, thanks a lot for your time (and your patience!)

Here are a few first comments. In addition, it's important we add tests on all the code we create. In this case, I suggest you to have a look at the Gandi tests: https://github.com/acmephp/acmephp/commit/9eeb20a1fd1ed164a76978d965b99ab3afaff5a9#diff-c141576f6cc9556a012d5180e3f6d1f45a3ee4da385787412bf01d6e2e7ea6d0

You should be able to do the same for OVH :)

**Pull request closed without merging**